internal/charmstore/resources.go

```diff
@@ -0,0 +1,255 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package charmstore // import "gopkg.in/juju/charmstore.v5-unstable/internal/charmstore"
+
+import (
+	"io"
+	"time"
+
+	"gopkg.in/errgo.v1"
+	"gopkg.in/juju/charm.v6-unstable"
+	"gopkg.in/juju/charmrepo.v2-unstable/csclient/params"
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+
+	"gopkg.in/juju/charmstore.v5-unstable/internal/mongodoc"
+)
+
+var resourceNotFound = errgo.Newf("resource not found")
+
+// ListResources returns the list of resources for the charm at the
+// latest revision for each resource.
+func (s Store) ListResources(entity *mongodoc.Entity, channel params.Channel) ([]*mongodoc.Resource, error) {
+	if entity.URL.Series == "bundle" {
+		return nil, errgo.Newf("bundles do not have resources")
+	}
+	if entity.CharmMeta == nil {
+		return nil, errgo.Newf("entity missing charm metadata")
+	}
+
+	var docs []*mongodoc.Resource
+	for name, meta := range entity.CharmMeta.Resources {
+		doc, err := s.latestResource(entity, channel, name)
+		if errgo.Cause(err) == resourceNotFound {
+			// Stick in a placeholder.
+			doc = &mongodoc.Resource{
+				CharmURL: entity.BaseURL,
+				Name:     meta.Name,
+				// Revision, Fingerprint, etc. are not set.
+			}
+		} else if err != nil {
+			return nil, errgo.Notef(err, "failed to get resource %q", name)
+		}
+		docs = append(docs, doc)
+	}
+	mongodoc.SortResources(docs)
+	return docs, nil
+}
+
+// ResourceBlob holds the information specific to a single resource blob.
+type ResourceBlob struct {
+	io.Reader
+
+	//Fingerprint is the SHA-384 checksum of the blob.
+	Fingerprint []byte
+
+	// Size is the size of the blob in bytes.
+	Size int64
+}
+
+// TODO(ericsnow) We will need Store.nextResourceRevision() to get the
+// value to pass to addResource().
+
+func (s Store) addResource(entity *mongodoc.Entity, name string, blob ResourceBlob, newRevision int) error {
+	if !charmHasResource(entity.CharmMeta, name) {
+		return errgo.Newf("charm does not have resource %q", name)
+	}
+
+	blobName, err := s.storeResource(blob)
+	if err != nil {
+		return errgo.Mask(err, errgo.Is(resourceNotFound))
+	}
+
+	doc := &mongodoc.Resource{
+		CharmURL:    entity.BaseURL,
+		Name:        name,
+		Revision:    newRevision,
+		Fingerprint: blob.Fingerprint,
+		Size:        blob.Size,
+		BlobName:    blobName,
+		UploadTime:  time.Now().UTC(),
+	}
+	if s.insertResource(doc); err != nil {
+		if err := s.BlobStore.Remove(blobName); err != nil {
+			logger.Errorf("cannot remove blob %s after error: %v", blobName, err)
+		}
+		return errgo.Mask(err)
+	}
+	return nil
+}
+
+func (s Store) insertResource(doc *mongodoc.Resource) error {
+	if err := doc.Validate(); err != nil {
+		return errgo.Mask(err)
+	}
+	err := s.DB.Resources().Insert(doc)
+	if err != nil && !mgo.IsDup(err) {
+		return errgo.Notef(err, "cannot insert resource")
+	}
+	// TODO(ericsnow) Also fail for dupe?
```

Comment:
My opinion: yes. Fail fast principle; if the caller is attempting to insert a duplicate, we're probably on a bad code path.